philc/vimium
# Safari web extension [WIP]

# Description
This creates a Safari web extension-wrapper around the current code, with the idea of being able to run Vimium natively in Safari (See #3610). I've basically just taken the steps from other users in that issue and wrapped it up in a PR for now, with the hope that we can collaborate on this here.

# Status
Right now this is runnable, but not everything is working, as in https://github.com/philc/vimium/issues/3610#issuecomment-884343691 some commands doesn't work (see below).

## How to test
1. Get the branch for this PR.
2. Open `Vimium/Vimium.xcodeproj` in XCode (You'll need XCode, I'm running XCode 12.5)
3. Run the project, then you'll need to make sure that you have enabled running unsigned extensions in safari: https://developer.apple.com/documentation/safariservices/safari_web_extensions/running_your_safari_web_extension
4. On further launches you'll just need to build in XCode, and it should reload properly.
5. For debugging, open `Develop>Web Extension Background Pages>Vimium` in Safari.

## Things to test
- [ ] Test through all known features, to make sure that they work as expected.
- [ ] Test on iPad/iOS

## Known bugs:
- [ ] Commands requiring the clipboard (`yy`, visual mode copy, `P`, etc.)
- [ ] `B` doesn't have access to bookmarks ([`bookmarks`-API isn't supported in safari](https://developer.mozilla.org/en-US/docs/Mozilla/Add-ons/WebExtensions/Browser_support_for_JavaScript_APIs#bookmarks))
- [ ] `F` loads a URL in the currently selected tab rather than a new one
- [ ] `X` doesn't seem to restore the last closed tab ([`sessions`-API isn't supported in safari](https://developer.mozilla.org/en-US/docs/Mozilla/Add-ons/WebExtensions/Browser_support_for_JavaScript_APIs#sessions))

# Questions/considerations
- [ ] Right now this PR adds a `Vimium` folder with an `.xcproj`, this is what was created by running `xcrun safari-web-extension-converter`, but might not be the desired folder structure
- [ ] Some code in `main.js` is running conditionally on whether or not we're on `chrome`, is this safe, or should we find a way to trigger it in similar conditions in safari?
- [ ] The `manifest.json` was changed to have a broader `matches` for one `.css` file, is this safe or do we need to do something smarter here?
- [ ] In #3610 there were some discussions on using other artwork, since the current logo has the Chrome logo in the background.
